compiler/rustc_parse/src/parser/path.rs

```diff
@@ -534,4 +537,51 @@ impl<'a> Parser<'a> {
         };
         Ok(Some(arg))
     }
+
+    fn get_ident_from_generic_arg(
+        &self,
+        gen_arg: Option<GenericArg>,
+        lo: Span,
+    ) -> PResult<'a, (Ident, Option<GenericArgs>)> {
+        if let Some(ref arg) = gen_arg {
+            match arg {
+                GenericArg::Type(t) => match &(*t).kind {
+                    ast::TyKind::Path(qself, ref path) => {
+                        if let Some(qself) = qself {
+                            let mut err = self.struct_span_err(
+                                qself.path_span,
+                                "invalid use of Self qualified paths",
+                            );
+                            err.span_label(
+                                qself.path_span,
+                                "not allowed in associated type constraints",
+                            );
+                            return Err(err);
+                        }
+                        if path.segments.len() == 1 {
+                            let path_seg = &path.segments[0];
+                            let ident = path_seg.ident;
+                            let gen_args = if let Some(ref gen_args) = path_seg.args {
+                                Some((**gen_args).clone())
+                            } else {
+                                None
+                            };
+
+                            return Ok((ident, gen_args));
+                        }
+                        let err = self.struct_span_err(
+                            path.span,
+                            "Expected a Path with a single PathSegment",
+                        );
+                        return Err(err);
+                    }
+                    _ => {}
+                },
+                _ => {}
+            }
+        }
+        let span = lo.to(self.prev_token.span);
+        let err = self.struct_span_err(span, "Expected identifier");
```

Comment:
This should be replaced with some more meaningful errors (I wrote some examples above).

Reply:
Could you also add test cases for `<A::B = C>`, `<1 = A>`, `<[u8; 1] = A>` and `<'a = A>`?